packages/expect/src/index.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * Object type O that includes non-nullable keys K
+ */
+export type Has<O extends {}, K extends string> = O &
+  Required<NonNullable<Pick<O, K & keyof O>>>;
+
+/**
+ * Object type O that includes exactly one non-nullable value among keys K
+ */
+export type HasOne<O extends {}, K extends string> = O &
+  {
+    [N in K]: Has<O, N>;
+  }[K];
+
+/**
+ * Asserts at runtime that `options` contains `key`
+ */
+export function has<O extends {}, K extends string>(
+  options: O,
+  key: K
+): asserts options is Has<O, K> {
+  // @ts-ignore to get around the fact that we know nothing about O
+  if (options[key] == null) {
+    throw new Error(`Expected parameter '${key}' not passed to function.`);
+  }
+}
+
+/**
+ * Asserts at runtime that `options` contains all `expectedKeys`
+ */
+export function options<O extends {}, K extends string>(
+  options: O,
+  expectedKeys: K[]
+): asserts options is Has<O, K> {
+  for (const key of expectedKeys) {
+    has(options, key);
+  }
+}
+
+/**
+ * Asserts at runtime that `options` contains at least one of `expectedKeys`
+ *
+ * Post-condition: this narrows type of `options` to include _exactly one_ of
+ * `expectedKeys`, even though at runtime this accepts more than one key.
+ */
+export function one<O extends {}, K extends string>(
```

Comment:
Why do we need `K extends string`? Couldn't you just type `expectedKeys` as something like `string[]`?

Reply:
I now expect the answer to this is similar to what we were talking about earlier. `K` provides more information to the TS compiler for later. Tell me if I'm wrong.

Reply:
It's necessary because we concern ourselves with actual key values, not just the arbitrary `string` type. So `one(options, ["size"])` becomes `one<O, "size">`, which means that later we can do `options.size` precisely.

Reply:
You're exactly right @eggplantzzz